### PR TITLE
FEA: engine accepts dpnp.ndarray and dpt.usm_ndarray objects as input data.

### PR DESCRIPTION
**Pull request by fcharras**
I think I'm happy with the result, the code is a little messy but not more than sklearn's and I think the PR basically succeeds in mimicking almost all aspects of sklearn's UX regarding what inputs are accepted and how it's casted under the hood, and also performance regarding the cautiousness with memory copies, while additionally creating support for `dpnp.ndarray` and `dpt.usm_ndarray` inputs, and recycling the most possible code that already exists in sklearn.

The PR still needs some polishing, and a few tests with `dpnp`/`dpt` tensors.

I think one (minor) difference is that, while sklearn try to convert numpy arrays with `object` dtype to `float64`, our engine will error out in this case because `dpt.asarray` refuses object dtype as input. Those kind of edge cases are complicated to unifiy because different array libraries can have different choices and I think it's fine we let it fail in this case.

### CODE REVIEW

File: benchmark/ext_helpers/daal4py.py [hunk 1]
```diff
@@ -37,7 +42,15 @@ def prepare_fit(self, X, y=None, sample_weight=None):
         return super().prepare_fit(X, y, sample_weight)
 
     def init_centroids(self, X):
-        return super(KMeansEngine, self).init_centroids(X)
```
Comment (fcharras) [inline benchmark/ext_helpers/daal4py.py:40]: This is a very important bugfix, I will open in a separate PR to fasttrack.

Basically, currently on main `k-means++` benchmark with the daal actually runs sklearn's. After fixing, we can see that the daal is more that 10 times faster than sklearn at kmeans++.

File: sklearn_numba_dpex/common/kernels.py [hunk 2]
```diff
@@ -304,6 +316,53 @@ def partial_sum_reduction(
         if first_work_id:
             add_first_cols(local_values, result, group_id)
 
+    return partial_sum_reduction
+
+
+@lru_cache
+def make_sum_reduction_2d_axis1_kernel(
+    size0, size1, work_group_size, device, dtype, fused_unary_func=None
```
Comment (fcharras) [inline sklearn_numba_dpex/common/kernels.py:324]: The changes to this sum are the addition of this arg `fused_unary_func` which enable fusing a unary ops on the elements of the input before summing. It's used in this PR to compute variance.

File: sklearn_numba_dpex/kmeans/tests/test_kmeans.py [hunk 3]
```diff
@@ -364,7 +366,7 @@ def _get_score_with_centers(centers):
         [
             -1827.22702,
             -1027.674243,
-            -865.257397,
+            -865.257501,
```
Comment (fcharras) [inline sklearn_numba_dpex/kmeans/tests/test_kmeans.py:369]: The difference in those values comes from numerical instability after including the X_mean removal step.

File: sklearn_numba_dpex/kmeans/engine.py [hunk 4]
```diff
@@ -156,155 +201,160 @@ def kmeans_single(self, X, sample_weight, centers_init):
...
+    # remove those hacks.
+    with sklearn.config_context(
+        array_api_dispatch=True,
+        assume_finite=True  # workaround 1: disable force_all_finite
+        # workaround 2: monkey patch get_namespace and _asarray_with_order to force
+        # dpctl.tensor array namespace
+    ), override_attr_context(
+        sklearn_validation,
+        get_namespace=_get_namespace,
+        _asarray_with_order=_asarray_with_order,
```
Comment (fcharras) [inline sklearn_numba_dpex/kmeans/engine.py:358]: Using monkey patching here enables a very efficient workflow as a whole. I wonder if rather than having those 2 functions limited to array api specs, we should also consider pluginying it.

File: benchmark/ext_helpers/daal4py.py (hunk 1 above)
Comment (ogrisel) [reply benchmark/ext_helpers/daal4py.py:40]: It would be very interesting to repriorize the optimization of kmeans++, both algorithmically (via subsampling as suggested by @jeremiedbb elsewhere) and with efficient Cython code for the computation of the argmin of distances if it's not already the case.

**Comment by fcharras** [comment]
Out of WIP

File: sklearn_numba_dpex/testing/config.py [hunk 5]
```diff
@@ -31,14 +31,15 @@ def override_attr_context(obj, **attrs):
...
+        for attr_name, attr_value in attrs.items():
+            # raise AttributeError if obj does not have the attribute attr_name
+            attrs_before[attr_name] = getattr(obj, attr_name)
+            setattr(obj, attr_name, attr_value)
+
+        yield
+
+    finally:
+        for attr_name, attr_value in attrs_before.items():
+            setattr(obj, attr_name, attr_value)
```
Comment (jjerphan) [inline sklearn_numba_dpex/testing/config.py:45]: How about this?
```suggestion
      attrs_before = dict()
      for attr_name, attr_value in attrs.items():
          attribute = getattr(obj, attr_name, None)
          if attribute is not None:
              # Only replace the value of the `attr_name` does exist.
              attrs_before[attr_name] = attribute
              setattr(obj, attr_name, attr_value)

      yield

      for attr_name, attr_value in attrs_before.items():
          setattr(obj, attr_name, attr_value)
```

File: sklearn_numba_dpex/kmeans/tests/test_kmeans.py [hunk 6]
```diff
@@ -336,8 +342,10 @@ def _get_score_with_centers(centers):
 
         kmeans.set_params(random_state=random_state)
         engine = KMeansEngine(kmeans)
-        engine.prepare_fit(X)
-        engine_kmeans_plusplus_centers = engine.init_centroids(X)
+        X_prepared, *_ = engine.prepare_fit(X)
+        engine_kmeans_plusplus_centers = engine.init_centroids(X_prepared)
+        engine_kmeans_plusplus_centers = dpt.asnumpy(engine_kmeans_plusplus_centers.T)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/tests/test_kmeans.py:347]: Do we need to `_t`-suffix one of them?

File: sklearn_numba_dpex/kmeans/tests/test_kmeans.py [hunk 7]
```diff
@@ -364,29 +372,36 @@ def _get_score_with_centers(centers):
...
-        random_state=random_state,
+        init="k-means++", n_clusters=n_clusters_sklearn_test, random_state=random_state
     )
     engine = KMeansEngine(estimator)
-    engine.prepare_fit(X, sample_weight=sample_weight)
-    centers, indices = engine._kmeans_plusplus(X)
+    X_prepared, *_ = engine.prepare_fit(X, sample_weight=sample_weight)
+
+    centers, indices = engine._kmeans_plusplus(X_prepared)
+    centers = dpt.asnumpy(centers.T)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/tests/test_kmeans.py:402]: Similarly, do we need to `_t`-suffix one of them?

File: sklearn_numba_dpex/common/kernels.py [hunk 8]
```diff
@@ -172,12 +171,16 @@ def make_sum_reduction_2d_axis1_kernel(size0, size1, work_group_size, device, dt
     minus_one_idx = np.int64(-1)
     two_as_a_long = np.int64(2)
 
-    is_1d = size1 is None
+    if fused_unary_func is None:
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:174]: Can you document `fused_unary_func`, please?

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 9]
```diff
@@ -401,6 +402,88 @@ def _relocate_empty_clusters(
...
+        X_t.shape[1],
+        device.max_work_group_size,
+        device=device,
+        dtype=compute_dtype,
+    )
+
+    X_mean = (sum_axis1_kernel(X_t) / compute_dtype(n_samples))[:, 0]
+
+    if (X_mean == 0).astype(int).sum() == len(X_mean):
+        X_mean = None
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:423]: Can you document why `None` is conventionally used in this case?

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 10]
```diff
@@ -401,6 +402,88 @@ def _relocate_empty_clusters(
...
+                n_clusters,
+                ops=_minus,
+                work_group_size=max_work_group_size,
+            )
+            broadcast_init_minus_X_mean(init, X_mean)
+
+    variance_kernel = make_sum_reduction_2d_axis1_kernel(
+        n_features * n_samples,
+        None,
+        max_work_group_size,
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:448]:
```suggestion
        size0=n_features * n_samples,
        size1=None,
        work_group_size=max_work_group_size,
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 11]
```diff
@@ -401,6 +402,88 @@ def _relocate_empty_clusters(
...
+    )
+    broadcast_X_plus_X_mean(X_t, X_mean)
+
+
+def _minus(x, y):
+    return x - y
+
+
+def _plus(x, y):
+    return x + y
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:484]: Can you group and document those helpers function for `ops`?

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 12]
```diff
@@ -663,4 +746,4 @@ def kmeans_plusplus(
         centers_t[:, c] = X_t[:, center_index]
         center_indices[c] = center_index
 
-    return centers_t.T, center_indices
+    return centers_t, center_indices
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:749]: What is the reason for this change (i.e. not transposing anymore)?

File: sklearn_numba_dpex/common/kernels.py [hunk 13]
```diff
@@ -100,6 +100,34 @@ def broadcast_division(dividend_array, divisor_vector):
...
+    value. lambda functions are advised against since the cache will not work with lamda
+    functions."""
+
+    global_size = math.ceil(size1 / work_group_size) * work_group_size
+    ops = dpex.func(ops)
+
+    # NB: inplace. # Optimized for C-contiguous array and for
+    # size1 >> preferred_work_group_size_multiple
+    @dpex.kernel
+    def broadcast_ops(left_operand_array, right_operand_vector):
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:117]: I think it's worth indicating that the left operand is modified in place and that the right one isn't modified at all.

**Review by jjerphan — COMMENTED** [review]
Thank you, @fcharras. Here is a first pass.

File: sklearn_numba_dpex/common/kernels.py (hunk 8 above)
Comment (fcharras) [reply sklearn_numba_dpex/common/kernels.py:174]: It's documented in the public function that follows. For readability it would be better I think to swap the definitions of those, but it would have increased the diff masked the true diff.

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 12 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/drivers.py:749]: "drivers" in drivers.py file now take and return the raw-est input and outputs. Pre and post processing regarding transposing, type, dtype, etc, that is required upstream or downstream is now located in the engine. I think those are better concepts for readability and maintaining.

File: sklearn_numba_dpex/testing/config.py (hunk 5 above)
Comment (fcharras) [reply sklearn_numba_dpex/testing/config.py:45]: The `try/finally` is not only about the `AttributeError` that was documented, but about any error that can be raised after `yield` and before exiting the context. I thought initially that before exiting the code after `yield` is always executed, in the same way it's true for  `__exit__` when expliciting context managers, but in case of such errors it seems not to be true, and caused failing tests (in the sklearn pipeline). Adding the `try/finally` fixes that.

About only replacing the value if `attr_name` does exist, i'd prefer not, raising an `AttributeError` is a safeguard against typos, or unseen changes in input objects, when using this context manager. Its only intended uses are for replacing attributes that exist and I like it to fail otherwise.

**Comment by fcharras** [comment]
Pushed your suggestions, also amended some changes to enable compute follow data. (the preferred device for compute is the device that stores de data if data is already on-device)

File: sklearn_numba_dpex/common/kernels.py [hunk 14]
```diff
@@ -304,6 +310,63 @@ def partial_sum_reduction(
...
+    sum_axis_size = size0 if n_rows is None else size1
+
+    # fused_unary_func is applied elementwise during the first pass on data, in the
+    # first kernel execution only.
+    fused_func_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
+        n_rows, work_group_size, fused_unary_func, dtype
+    )
+    # subsequent kernel calls only sum the data.
+    nofunc_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
+        n_rows, work_group_size, None, dtype
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/kernels.py:367]:
```suggestion
        n_rows, work_group_size, fused_unary_func=None, dtype=dtype
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 15]
```diff
@@ -401,6 +404,80 @@ def _relocate_empty_clusters(
...
+        X_t.shape[0],
+        X_t.shape[1],
+        device.max_work_group_size,
+        device=device,
+        dtype=compute_dtype,
+    )
+
+    X_mean = (sum_axis1_kernel(X_t) / compute_dtype(n_samples))[:, 0]
+
+    if (X_mean == 0).astype(int).sum() == len(X_mean):
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:424]:
```suggestion
    X_mean_is_zeroed = (X_mean == 0).astype(int).sum() == len(X_mean)

    if X_mean_is_zeroed:
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 16]
```diff
@@ -401,6 +404,80 @@ def _relocate_empty_clusters(
     )
 
 
+def prepare_data_for_lloyd(X_t, init, tol, copy_x):
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:407]: Can you document `prepare_data_for_lloyd` indicating that this is centering `X` for numerical stability?

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 17]
```diff
@@ -401,6 +404,80 @@ def _relocate_empty_clusters(
...
+        dtype=compute_dtype,
+        fused_unary_func=_square,
+    )
+    variance = variance_kernel(dpt.reshape(X_t, -1)) / n_features
+    tol = variance * tol
+
+    return X_t, X_mean, init, tol
+
+
+def restore_data_after_lloyd(X_t, X_mean):
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:465]: Similarly, can you document `restore_data_after_lloyd` indicating that this is translating `X` to its original position?

File: sklearn_numba_dpex/kmeans/engine.py [hunk 18]
```diff
@@ -47,101 +63,134 @@ class KMeansEngine(KMeansCythonEngine):
...
+        elif isinstance(init, str) and init == "k-means++":
+            centers_t, _ = self._kmeans_plusplus(X)
+
+        elif callable(init):
+            centers = init(X, self.estimator.n_clusters, random_state=self.random_state)
+            centers_t = self._check_init(centers, X)
+
+        else:
+            # NB: sampling without replacement must be executed sequentially so
+            # it's better done on CPU
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:161]: :+1:

File: sklearn_numba_dpex/kmeans/engine.py [hunk 19]
```diff
@@ -47,101 +63,134 @@ class KMeansEngine(KMeansCythonEngine):
...
+            centers_idx = self.random_state.choice(
+                X.shape[0], size=n_clusters, replace=False
+            )
+            # Poor man's fancy indexing
+            # TODO: write a kernel ? or replace with better equivalent when available ?
+            centers_t = dpt.concat(
+                [dpt.expand_dims(X[center_idx], axes=1) for center_idx in centers_idx],
+                axis=1,
+            )
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:170]: Should we open an issue on `dpctl`? What do you think?

File: sklearn_numba_dpex/kmeans/engine.py [hunk 20]
```diff
@@ -47,101 +63,134 @@ class KMeansEngine(KMeansCythonEngine):
...
         )
+        return centers_t, center_indices
+
+    def kmeans_single(self, X, sample_weight, centers_init_t):
+        # ???: using `.all()` often segfaults
+        # TODO: minimal reproducer and issue at dpnp
+        # or write a kernel ?
+        use_uniform_weights = (sample_weight == sample_weight[0]).astype(
+            int
+        ).sum() == len(sample_weight)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:188]:
```suggestion
        use_uniform_weights = (
            (sample_weight == sample_weight[0]).astype(int).sum()
            == len(sample_weight)
        )
```

File: sklearn_numba_dpex/kmeans/engine.py [hunk 21]
```diff
@@ -156,155 +205,170 @@ def kmeans_single(self, X, sample_weight, centers_init):
...
         )
 
+    def prepare_prediction(self, X, sample_weight):
+        X = self._validate_data(X, reset=False)
+        sample_weight = self._check_sample_weight(sample_weight, X)
+        return X, sample_weight
+
     def get_labels(self, X, sample_weight):
-        labels, _ = self._get_labels_inertia(X, with_inertia=False)
+        # TODO: sample_weight actually not used for get_labels. Fix in sklearn ?
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:223]: Is there an issue already open for it?

File: sklearn_numba_dpex/kmeans/engine.py [hunk 22]
```diff
@@ -156,155 +205,170 @@ def kmeans_single(self, X, sample_weight, centers_init):
...
-            text = (
-                f"KMeans is set to compute with type {compute_dtype} but this type is "
-                f"not supported by the device {self.device.name}. The computations "
-                f"will default back to float32 type."
-            )
-            compute_dtype = np.float32
-
+            sample_weight = dpt.ones(n_samples, dtype=dtype, device=device)
+        elif isinstance(sample_weight, numbers.Number):
+            sample_weight = dpt.full(n_samples, 1, dtype=dtype, device=device)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:306]: What are the reasons not to have only one branch here?

**Review by jjerphan — APPROVED** [review]
LGTM up to a few suggestions and questions.

Thank you, @fcharras!

File: sklearn_numba_dpex/kmeans/engine.py (hunk 22 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:306]: A mistake! the latter line should read:

```suggestion
            sample_weight = dpt.ones(n_samples, dtype=dtype, device=device)
        elif isinstance(sample_weight, numbers.Number):
            sample_weight = dpt.full(n_samples, sample_weight, dtype=dtype, device=device)
```

File: sklearn_numba_dpex/kmeans/engine.py [hunk 23]
```diff
@@ -47,101 +62,134 @@ class KMeansEngine(KMeansCythonEngine):
 
     """
 
+    # This class attribute can alter globally the attributes `device` and `order` of
+    # future instances. It is only used for testing purposes, using
+    # `sklearn_numba_dpex.testing.config.override_attr_context` context, for instance
+    # in the benchmark script.
+    # For normal usage, the compute will follow the __compute_follow_data__ principle.
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:70]: Yes, let me replace with `*compute follow data*` which would be more correct makrdown.

The concept is anchored in dpctl/numba_dpex desiogn

- in dpctl [1](https://intelpython.github.io/dpctl/latest/docfiles/dpctl/dpctl.utils_pyapi.html) 2

- in numba dpex [1](https://intelpython.github.io/numba-dpex/latest/user_guides/kernel_programming_guide/selecting_device.html#defining-the-execution-queue-for-a-kernel-function) [2](https://github.com/IntelPython/numba-dpex/issues/816#issuecomment-1323080669)

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 16 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/drivers.py:407]: :heavy_check_mark: mostly reported [sklearn docstring](https://github.com/scikit-learn/scikit-learn/blob/17df37aee/sklearn/cluster/_kmeans.py#L332)

File: sklearn_numba_dpex/kmeans/engine.py (hunk 19 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:170]: `dpctl` implements the Array API and the Array API doesn't list such tools. Best things to have would be `take` but here are the list of supported functions [1](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html) [2](https://data-apis.org/array-api/latest/API_specification/indexing.html)

I'm not aware enough of where the array api is going to say what would be the best action here :thinking:

File: sklearn_numba_dpex/kmeans/engine.py (hunk 21 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:223]: Done in https://github.com/scikit-learn/scikit-learn/issues/25066

File: sklearn_numba_dpex/kmeans/engine.py (hunk 19 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:170]: Actually `take` has been added recently to the Array APi, [dpctl issue](https://github.com/IntelPython/dpctl/issues/1003)

**Comment by fcharras** [comment]
Thanks for the review @jjerphan I've pushed your suggestions and answered other comments. I'll go ahead and merge when the pipeline is green.
